packages/grid/x-data-grid/src/hooks/features/columns/useGridColumns.tsx

```diff
@@ -347,6 +348,22 @@ export function useGridColumns(
     [props.components.ColumnsPanel, props.componentsProps?.columnsPanel],
   );
 
+  const addColumnMenuItems = React.useCallback<GridPipeProcessor<'columnMenu'>>(
+    (columnMenuItems, column) => {
+      if (props.disableColumnSelector) {
+        return columnMenuItems;
+      }
+
+      if (column.hideable === false) {
+        return [...columnMenuItems, 'ColumnMenuColumnsItem'];
+      }
+
+      return [...columnMenuItems, 'ColumnMenuColumnsItem', 'ColumnMenuHideItem'];
```

Comment:
Seems like this logic is duplicated here
https://github.com/MBilalShafi/mui-x/blob/f59b06645241a52710e11ecab8ac846c7adf3755/packages/grid/x-data-grid/src/components/menu/columnMenu/default/GridColumnMenuHideItem.tsx#L57-L63
and in a few other components.
Maybe keep it here and remove it from the components?

Reply:
I kept it in components because sometimes user may use these items without using our Column Menu component, like:
```tsx
const CustomColumnMenu = (props) => {
  return (
    <Stack>
      <ColumnMenuHideItem />
      {...otherItems}
    </Stack>
  )
}
```

In these cases too, if a certain item shouldn't be rendered due to some prop etc it should behave as expected. Right?

Reply:
Right, that makes sense.
Maybe keep the logic in menu components and remove it from `useGridColumns` then?

Reply:
The idea is to control the rendering of default components from respective hooks based on different conditions, if we remove this logic from here, it might break this convention. But it also makes sense not to have duplicate checks, I'll remove the checks from hooks on the components where we have the same logic internally 👍

Reply:
I'm convinced :+1:

Reply:
> I think it's not necessary for the menu items of the default variant.

Do you think the following scenario is not valid for default variant?

_Let's say, if the user provides a custom component (override) for one of these components, they will not work as expected (e.g ColumnMenuColumnItem will still be rendered even if props.disableColumnSelector is true)_

> Side node: we need to allow to remove menu items like in the old approach. Maybe remove the item from the array if displayOrder: -1.

Right now they can do something like:

```tsx
<GridColumnMenuDefault
  components={{
    // remove sort item
    ColumnMenuSortItem: null
  }}
/>
```

Reply:
> The scenario you gave would be covered by having the check inside the pre-processor, like you already implemented. No matter which component the user provides to the slot, the menu item will not be rendered because ColumnMenuSortItem is not added to the array by the pre-processor. I only think we can remove the 2nd check from inside the component, for the default variant.

Sorry about the confusion, I meant this usecase when user doesn't use our column menu component but overrides the `ColumnMenu` slot of `DataGrid`

```tsx
<DataGrid
  components={{ ColumnMenu: UserColumnMenu }}
/>
```

And CustomColumnMenu could be implemented this way:

```tsx
const CustomColumnMenu = (props) => {
  return (
    <Stack>
      {/* Possibly use <ColumnMenuContainer /> */}
      <ColumnMenuColumnsItem {...props} />
      {...otherItems}
    </Stack>
  )
}
```

Now, in this scenario, pipe processor will not come into action and if user sets `props.disableColumnSelector` to `true` it will not hide `ColumnMenuColumnsItem`